lib/fizzy/parser_expr.cpp

```diff
@@ -88,50 +83,55 @@ parser_result<std::optional<ValType>> parse_blocktype(const uint8_t* pos, const
     return {validate_valtype(type), pos};
 }
 
-void update_stack_height(ControlFrame& frame, int stack_height_required, int stack_height_change)
+void update_operand_stack(const ControlFrame& frame, Stack<ValType>& operand_stack,
+    span<const ValType> inputs, span<const ValType> outputs)
 {
-    if (frame.stack_height < frame.parent_stack_height + stack_height_required)
-    {
-        // Stack is polymorphic after unreachable instruction: underflow is ignored,
-        // but we need to count stack growth to detect extra values at the end of the block.
-        if (frame.unreachable)
-        {
-            // Add instruction's/call's output values only.
-            frame.stack_height =
-                frame.parent_stack_height + stack_height_required + stack_height_change;
-        }
-        else
-            throw validation_error{"stack underflow"};
-    }
-    else
+    const auto frame_stack_height = static_cast<int>(operand_stack.size());
+    const auto inputs_size = static_cast<int>(inputs.size());
+
+    // Stack is polymorphic after unreachable instruction: underflow is ignored,
+    // but we need to count stack growth to detect extra values at the end of the block.
+    if (!frame.unreachable && frame_stack_height < frame.parent_stack_height + inputs_size)
+        throw validation_error{"stack underflow"};
+
+    // Update operand_stack.
+    for (auto it = inputs.begin() + inputs.size(); it != inputs.begin(); --it)
```

Comment:
Would be nice to have reverse iterator for `span` instead of this, but I would leave it for another PR.

Reply:
If we go ahead with this PR I'd prefer span to be merged separately anyway and in that PR this could be implemented too.